components/script/fetch.rs

```diff
@@ -266,6 +273,201 @@ pub(crate) fn Fetch(
+            let activate_after = *activate_after as u64;
+            if elapsed < activate_after {
+                let sleep_duration = activate_after - elapsed;
+                thread::sleep(Duration::from_millis(sleep_duration));
+            }
+            // Step 6.2. Process deferredRecord.
+            deferred_record_clone.lock().unwrap().process();
+
+            // Last step of https://fetch.spec.whatwg.org/#process-a-deferred-fetch
+            //
+            // Step 4. Queue a global task on the deferred fetch task source with
+            // deferredRecord’s request’s client’s global object to run deferredRecord’s notify invoked.
+            trusted_global
+                .root()
+                .task_manager()
+                .deferred_fetch_task_source()
+                .queue(task!(notify_deferred_record: move || {
+                    deferred_record_clone.lock().unwrap().activate();
+                }));
+        }),
+    );
+    // Step 7. Return deferredRecord.
+    deferred_record
+}
+
+/// <https://fetch.spec.whatwg.org/#dom-window-fetchlater>
+#[allow(non_snake_case, unsafe_code)]
+pub(crate) fn FetchLater(
+    window: &Window,
+    input: RequestInfo,
+    init: RootedTraceableBox<DeferredRequestInit>,
+    can_gc: CanGc,
+) -> Fallible<DomRoot<FetchLaterResult>> {
+    let global_scope = window.upcast();
+    // Step 1. Let requestObject be the result of invoking the initial value
+    // of Request as constructor with input and init as arguments.
+    let request_object = Request::constructor(global_scope, None, can_gc, input, &init.parent)?;
+    // Step 2. If requestObject’s signal is aborted, then throw signal’s abort reason.
+    let signal = request_object.Signal();
+    if signal.aborted() {
+        // TODO: Throw abort reason here without also returning an `Err`
+        //
+        // let cx = GlobalScope::get_cx();
+        // rooted!(in(*cx) let mut abort_reason = UndefinedValue());
+        // signal.Reason(cx, abort_reason.handle_mut());
+        // let ar = enter_realm(window);
+        // unsafe {
+        //     assert!(!JS_IsExceptionPending(*cx));
+        //     JS_SetPendingException(
+        //         *cx,
+        //         abort_reason.handle().into(),
+        //         ExceptionStackBehavior::Capture,
+        //     );
+        //     report_pending_exception(cx, true, InRealm::Entered(&ar), can_gc);
+        // }
+        return Err(Error::Abort);
+    }
+    // Step 3. Let request be requestObject’s request.
+    let request = request_object.get_request();
+    // Step 4. Let activateAfter be null.
+    let mut activate_after = Finite::wrap(0_f64);
+    // Step 5. If init is given and init["activateAfter"] exists, then set
+    // activateAfter to init["activateAfter"].
+    if let Some(init_activate_after) = init.activateAfter.as_ref() {
+        activate_after = *init_activate_after;
+    }
+    // Step 6. If activateAfter is less than 0, then throw a RangeError.
```

Comment:
The spec doesn't consider when `activate_after` is still null. I think this is a spec bug, but haven't figured out yet if that's the case. The tests seem to imply that though.